### ISSUE TITLE
 tf.contrib.learn.monitors.ValidationMonitor hangs when passed input_fn parameter

### ISSUE THREAD

**Issue by krautt** [tensorflow/tensorflow#6727]
Hello!
I have been working my way through the  tf.contrib.learn [tutorials](https://www.tensorflow.org/tutorials/) and have been attempting to integrate the tf.contrib.learn.monitors.ValidationMonitor into the 'deep' classifier in wide_n_deep.py as shown below.

### Environment info
Operating System:  I ran this on Ubuntu Server 16.04 on a physical I7 with a GTX1080 gpu when i noticed the problem.  I know that i was using the GPU on the original physical box from previous tests, and because during the hang the nvidia_smi command showed considerable load on the GPU.  I was able to replicate the problem with CPU on a 16.04 VM as well.

Installed version of CUDA and cuDNN: 
```
/home/andersonjas/libcudnn5-dev_5.1.5-1+cuda8.0_amd64.deb
/home/andersonjas/libcudnn5_5.1.5-1+cuda8.0_amd64.deb
```

If installed from binary pip package, provide:

1. A link to the pip package you installed:
from Anaconda 2.7 64 bit package:
```
pip install tensorflow
```
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.

```
andersonjas@ubuntu:~$ python -c "import tensorflow; print(tensorflow.__version__)"
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcudnn.so.5 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcurand.so locally
0.11.head

```
If installed from source, provide 

1. The commit hash (`git rev-parse HEAD`)
2. The output of `bazel version`

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

```python
validation_monitor = tf.contrib.learn.monitors.ValidationMonitor(input_fn=lambda:input_fn(df_test), 
                       every_n_steps=50)
m.fit(input_fn=lambda: input_fn(df_train), steps=151,monitors=[validation_monitor])
```

Doing this in a jupyter notebool causes the code to hang indefinitely.  To make completely sure that i don't have a bug in my own code i can make the following change:

```python
validation_monitor = tf.contrib.learn.monitors.ValidationMonitor(input_fn=lambda:input_fn(df_test), 
                       every_n_steps=50)
m.fit(input_fn=lambda: input_fn(df_train), steps=151) #,monitors=[validation_monitor])
```

and then the code executes fine.

### What other attempted solutions have you tried?
I also built an input_fn interface to the iris and boston housing price predictor code code, each showed similar 'hangs'

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).

As a noob, i'm learning that esoteric error messages are a luxury :-),  in this case the code just hangs indefinitely.

**Comment by krautt** [comment]
FWIW,  tf.contrib.learn.monitors.ValidationMonitor  works like a champ when passed x,y parameters instead of input_fn, as instructed by the [tutorial](https://www.tensorflow.org/tutorials/monitors/)

**Comment by JaySWang** [comment]
suffered from the same issue

**Comment by prb12** [comment]
@martinwicke Who would be the right person to look at ValidationMonitor now?  (@ilblackdragon wrote this)

@JaySWang why did you unassign him?

**Comment by vmalyi** [comment]
I have the same issue. Please see details below.

### Environment info
Operating System:
- MacOS Sierra 10.12.3 (16D32)
- Python 2.7.10

**Installed version of CUDA and cuDNN:** 
n/a

**A link to the pip package you installed:**
https://pypi.python.org/packages/db/0b/8ca3299122cb22f763df7076938b12d7f9e8a814d3ab7a80df6a41ef81f7/tensorflow-1.0.1-cp27-cp27m-macosx_10_11_x86_64.whl#md5=5f0f52fc2e5243d40d6ced52c6a56c69
**The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.**
1.0.1

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code):
[linear.py.zip](https://github.com/tensorflow/tensorflow/files/837779/linear.py.zip)

### What other attempted solutions have you tried?
n/a

### Logs or other output that would be helpful
[6727_logs.txt](https://github.com/tensorflow/tensorflow/files/837781/6727_logs.txt)

**Comment by gmacleod** [comment]
I just found a workaround to what might be the same issue in Tensorflow 0.12.1. Maybe it will be helpful here.

ValidationMonitor did not support the x=...y=... input, and when using input_fn, I found it would loop forever the first time it tried to run the monitor during a fit() call. The log output looked like it was calculating the validation metrics over and over in an infinite loop.

After looking at the source for ValidationMonitor I found a parameter eval_steps that defaulted to None. In other places in the contrib.learn codebase 'None' implies 'forever', so I tried adding eval_steps=1 to the constructor, and now it works as advertised.

e.g.
```
        validation_monitor = tf.contrib.learn.monitors.ValidationMonitor(
            input_fn=lambda: input_function(test_data, test_labels),
            eval_steps=1,  # Try adding this
            metrics=validation_metrics,
            every_n_steps=10,
        )   

        estimator.fit(
            input_fn=lambda: input_function(training_data, training_labels),
            steps=training_epochs,
            monitors=[validation_monitor],
        ) 
```

Hope that helps!

**Comment by vmalyi** [comment]
Hi @gmacleod,

thanks for pointing out this ** eval_steps** parameter. 

I can confirm that the issue can't be reproduced once eval_steps set to "1".

**Comment by martinwicke** [comment]
I will close this issue. The semantics might be awkward, but the workaround is simple and we have deprecated monitors in favor of Hooks.

**Comment by krautt** [comment]
Awesome. Thanks for the follow up.  Eager to try it out.

**Comment by hehongliang** [comment]
Great, it solved my problem!

**Comment by burness** [comment]
@gmacleod  Thanks for pointing out this solution! It happened in R1.1 too

**Comment by GuohongLi** [comment]
@gmacleod In tf1.1.0 , even adding eval_steps=1 to the constructor, it just output the validation log one time at the end of first n_steps, not every n_steps. Do you have the same problem?

**Comment by ikarosilva** [comment]
@gmacleod Thanks!! Happens in R1.1 as well.

**Comment by burness** [comment]
@GuohongLi  Maybe you should change the params in  `config=tf.contrib.learn.RunConfig()` to save the model weights, and then you can make the monitors. [](https://www.tensorflow.org/get_started/monitors)

**Comment by lancerts** [comment]
This indeed solves my problem. However, now the validationmonitor only trigger **ONE** time, any solution for this?


 validation_monitor = tf.contrib.learn.monitors.ValidationMonitor(
    input_fn=lambda: input_fn(df_test),
    every_n_steps=25,metrics=validation_metrics,
    eval_steps=1,
    early_stopping_metric="auc",
    early_stopping_metric_minimize=False,
    early_stopping_rounds=100)

INFO:tensorflow:Validation (step 25): loss = 404.011, accuracy = 0.859006, labels/prediction_mean = 1.88953e-07, labels/actual_label_mean = 0.140994, accuracy/baseline_label_mean = 0.140994, auc = 0.5, auc_precision_recall = 0.570497, accuracy/threshold_0.500000_mean = 0.859006, precision/positive_threshold_0.500000_mean = 0.0, recall/positive_threshold_0.500000_mean = 0.0, precision = 0.0, recall = 0.0, global_step = 1
INFO:tensorflow:global_step/sec: 5.23263
INFO:tensorflow:loss = 51.8876, step = 101 (19.112 sec)
INFO:tensorflow:global_step/sec: 28.5071
INFO:tensorflow:loss = 5.04256, step = 201 (3.508 sec)
INFO:tensorflow:global_step/sec: 25.6652
INFO:tensorflow:loss = 11.7246, step = 301 (3.896 sec)
INFO:tensorflow:global_step/sec: 29.4536
INFO:tensorflow:loss = 10.9838, step = 401 (3.395 sec)
INFO:tensorflow:global_step/sec: 28.0369
INFO:tensorflow:loss = 8.74781, step = 501 (3.567 sec)
INFO:tensorflow:global_step/sec: 27.9563
INFO:tensorflow:loss = 10.7869, step = 601 (3.577 sec)
INFO:tensorflow:global_step/sec: 27.0742
INFO:tensorflow:loss = 17.0162, step = 701 (3.694 sec)
INFO:tensorflow:global_step/sec: 27.3682
INFO:tensorflow:loss = 4.63466, step = 801 (3.654 sec)
INFO:tensorflow:global_step/sec: 27.6004
INFO:tensorflow:loss = 6.31007, step = 901 (3.623 sec)
INFO:tensorflow:global_step/sec: 27.6511
INFO:tensorflow:loss = 2.64077, step = 1001 (3.616 sec)
INFO:tensorflow:global_step/sec: 27.4551
INFO:tensorflow:loss = 1.91578, step = 1101 (3.642 sec)
INFO:tensorflow:global_step/sec: 27.7833
INFO:tensorflow:loss = 7.05945, step = 1201 (3.599 sec)
INFO:tensorflow:global_step/sec: 28.0504
INFO:tensorflow:loss = 2.11703, step = 1301 (3.565 sec)
INFO:tensorflow:global_step/sec: 28.0631
INFO:tensorflow:loss = 7.89242, step = 1401 (3.563 sec)
INFO:tensorflow:global_step/sec: 27.6421
INFO:tensorflow:loss = 13.6624, step = 1501 (3.618 sec)
INFO:tensorflow:global_step/sec: 27.7735
INFO:tensorflow:loss = 2.61309, step = 1601 (3.601 sec)
INFO:tensorflow:global_step/sec: 27.5637
INFO:tensorflow:loss = 7.0057, step = 1701 (3.628 sec)
INFO:tensorflow:global_step/sec: 27.2692
INFO:tensorflow:loss = 5.1032, step = 1801 (3.667 sec)
INFO:tensorflow:global_step/sec: 27.225
INFO:tensorflow:loss = 1.73602, step = 1901 (3.673 sec)
INFO:tensorflow:Saving checkpoints for 2000 into /var/folders/dg/jxsw2s955c1f91h17nxqm4hm395n3_/T/tmpquiqiew6/model.ckpt.
INFO:tensorflow:Loss for final step: 1.56839.

**Comment by sethtroisi** [comment]
@gmacleod Thanks eval_steps = 1 fixed the problem!

@lancerts did you find a solution to validation only triggering once?

**Comment by lancerts** [comment]
from tensorflow output,

INFO:tensorflow:Saving dict for global step 39807: accuracy = 0.85421, accuracy/baseline_label_mean = 0.14821, accuracy/threshold_0.500000_mean = 0.85421, auc = 0.686321, global_step = 39807, labels/actual_label_mean = 0.14821, labels/prediction_mean = 0.146081, loss = 0.39175, precision/positive_threshold_0.500000_mean = 0.580026, recall/positive_threshold_0.500000_mean = 0.0591728, validate_confusion_matrix = [[84052 699]
[14430 819]], validate_streaing_precision = 0.580026, validate_streaing_recall = 0.0591728, validate_streaming_auc = 0.525859
WARNING:tensorflow:Skipping summary for validate_confusion_matrix, must be a float or np.float32.
WARNING:tensorflow:Skipping summary for global_step, must be a float or np.float32.
INFO:tensorflow:Validation (step 40000): loss = 0.39175, accuracy = 0.85421, labels/prediction_mean = 0.146081, labels/actual_label_mean = 0.14821, accuracy/baseline_label_mean = 0.14821, auc = 0.686321, accuracy/threshold_0.500000_mean = 0.85421, precision/positive_threshold_0.500000_mean = 0.580026, recall/positive_threshold_0.500000_mean = 0.0591728, validate_confusion_matrix = [[84052 699]
[14430 819]], validate_streaing_precision = 0.580026, validate_streaing_recall = 0.0591728, validate_streaming_auc = 0.525859, global_step = 39807

Why are this two AUC differs so much? When i use model.evaluate on train, validate, test data set, all the output AUCs are very close to the first auc shows in bold above.
So what is the validation_streaming_auc is calculating? I have reset the local variables in each epoch.
